cat/spirv.cat

```diff
@@ -57,10 +57,11 @@ let avvisinc = ((avvisSem | avvisSem^-1) \ id) |
 let posctosem = ((SC0 * SEMSC0) | (SC1 * SEMSC1)) & po
 let posemtosc = ((SEMSC0 * SC0) | (SEMSC1 * SC1)) & po
 
-let asmo = co & (ATOM * ATOM)
+let asmo = co & ((ATOM | IW) * ATOM)
 // From-read is defined based on scoped modification order
-// instead of coherence
-let fr = rf^-1; asmo
+// instead of coherence. For reads that read-from init, they
+// from-read all writes at the same location.
+let fr = (rf^-1; asmo) | (([IW]; rf)^-1; co)
```

Comment:
Can't you define `let asmo = co & (ATOM * ATOM | IW * W)`? Then you don't need to change `fr`.

Reply:
I don't think that matches the specification

> For a given atomic write A, all atomic writes that are mutually-ordered with A occur in an order known as A’s scoped modification order. A’s scoped modification order relates no other operations.

Precisely, the "relates no other operations" part.

Even if it were, I still need to redefine `fr`. Also, this way [is closer to the alloy model.](https://github.com/KhronosGroup/Vulkan-MemoryModel/blob/66c31d9cfcc03b1f5231779fbf3c42dc983b62e2/alloy/spirv.als#L308-L312)

Reply:
By "not change `fr`" I meant that you can keep `fr = rf^-1;asmo` not the original one of course.
However, I agree with not having non-atomic writes in `asmo`. Do you still think that `IW * ATOM` should be there, i.e., should initial writes be considered atomic or not? They probably are, but then we should maybe just define `let ATOM = ATOM | IW` to make init writes appear atomic everywhere?

Lastly, why don't we define `let fr = (rf^-1; asmo) | (([IW]; rf)^-1; loc;[W] \ id)` to really match the alloy model? It also avoids this awkward usage of `co`.
It could also be refactored to `let fr = rf^-1;(asmo | ([IW];loc;[W] \ id))` which I think is nicer, but the above version is a little closer to the original alloy model.

Reply:
> They probably are, but then we should maybe just define let ATOM = ATOM | IW to make init writes appear atomic everywhere?

I think they probably are there, but only for those variables having at least once atomic access. If we use `ATOM = ATOM | IW`, we would be even adding initial stores for variables that are only accessed as non-atomic. Using `asmo = co & ((ATOM | IW) * ATOM)` we don't have this problem.

> Lastly, why don't we define let fr = (rf^-1; asmo) | (([IW]; rf)^-1; loc;[W] \ id) to really match the alloy model?

That might be the best. I wrote the other version manually simply following the "For reads that read-from init, they from-read all writes at the same location."